backend/core/modules/auth/service.go

```diff
@@ -0,0 +1,141 @@
+	ctx          context.Context
+	tvaultPath   string
+	databasePath string
+	databaseKey  []byte
+	isUnlocked   bool
+}
+
+func NewService(ctx context.Context) Service {
+	return &service{
+		ctx:          ctx,
+		tvaultPath:   authutils.GetTVaultPath(),
+		databasePath: authutils.GetDatabasePath(),
+		isUnlocked:   false,
+	}
+}
+
+func (s *service) Initialize(ctx context.Context) error {
+	s.ctx = ctx
+
+	// create directory if they don't exists
+	vaultDir := filepath.Dir(s.tvaultPath)
+	if err := os.MkdirAll(vaultDir, 0755); err != nil {
+		return fmt.Errorf("failed to create vault directory: %w", err)
+	}
+
+	// create tmp directory for decrypted files
+	tempDir := authutils.GetTempDir()
+	if err := os.MkdirAll(tempDir, 0755); err != nil {
+		return fmt.Errorf("failed to create temp directory: %w", err)
+	}
+
+	runtime.LogInfo(ctx, "Auth service initialized")
+	return nil
+}
+
+func (s *service) IsFirstTimeSetup() bool {
+	// Check if the tvault file exists
+	_, err := os.Stat(s.tvaultPath)
+	return os.IsNotExist(err)
+}
+
+func (s *service) CreatePassword(password string) error {
+	if len(password) < 6 {
+		return constants.ErrPasswordTooShort
+	}
+
+	//generate random database key | TODO: move this outside of this function
+	dbKey := make([]byte, constants.KeyLength)
+	if _, err := rand.Read(dbKey); err != nil {
+		return fmt.Errorf("failed to generate database key: %w", err)
+	}
+
+	//generate random salt
+	salt := make([]byte, constants.SaltLength)
+	if _, err := rand.Read(salt); err != nil {
+		return fmt.Errorf("failed to generate salt: %w", err)
+	}
+
+	config := argon2.MemoryConstrainedDefaults()
+
+	raw, err := config.HashRaw([]byte(password))
+	if err != nil {
+		return fmt.Errorf("failed to hash password: %w", err)
+	}
+
+	encryptedDBKey, err := authutils.EncryptData(dbKey, raw.Hash)
+	if err != nil {
+		argon2.SecureZeroMemory(raw.Hash)
+		return fmt.Errorf("failed to encrypt database key: %w", err)
+	}
+
+	if err := authutils.InitializeTVaultHeader(raw.Salt, encryptedDBKey); err != nil {
+		argon2.SecureZeroMemory(raw.Hash)
+		return fmt.Errorf("failed to initialize tvault header: %w", err)
+	}
+
+	argon2.SecureZeroMemory(raw.Hash)
+
+	// Store database key in memory
+	s.databaseKey = dbKey
+	s.isUnlocked = true
+
+	runtime.LogInfo(s.ctx, "Password created successfully")
+	return nil
+}
+
+func (s *service) DecryptDatabaseKey(password string) error {
+	runtime.LogInfo(s.ctx, "Verifying password")
+
+	salt, encryptedDBKey, err := authutils.ReadTVaultHeader()
+	if err != nil {
+		return err
+	}
+
+	config := argon2.MemoryConstrainedDefaults()
+
+	raw, err := config.Hash([]byte(password), salt)
+	if err != nil {
+		return fmt.Errorf("failed to derive key: %w", err)
+	}
+
+	dbKey, err := authutils.DecryptData(encryptedDBKey, raw.Hash)
+	if err != nil {
+		argon2.SecureZeroMemory(raw.Hash)
+		runtime.LogInfo(s.ctx, "Invalid password")
+		return constants.ErrInvalidPassword
+	}
+
+	argon2.SecureZeroMemory(raw.Hash)
+
```

Comment:
this is just an audit optimization: 

i think if you put all operations relating to cryptographic routines in the same file and only perform cryptographic operations in that one place (calling into it from e.g. `auth/service.go`) then i think that will give the auditors an easier time when it comes to analyzing faults and mistakes :)

Reply:
so like the actual lines of code calling argon2, aes etc